### ISSUE TITLE
Return type of functions intersection incrorrect

### ISSUE THREAD

**Issue by caikan** [microsoft/TypeScript#54221]
# Bug Report

### 🔎 Search Terms

function intersection

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEBhEUIEYkG8CwAUEkgJbAAuEATgGYLIBiwArmAEbW4AUAlAFxIW7agF9CxMpVr0kDKOSpkA5t35J5i4ErFFSFanUSzgUVQKEcqAHw3Kd4pOQCeMRkgC8x0wDJjrSw40zMDw5CQA9sCO0ORcNAIMagBKEOTMVMAAKi4QADwMAHy4DsRUqelRNLwOOiJAA)

### 💻 Code

```ts
// We can quickly address your report if:
//  - The code sample is short. Nearly all TypeScript bugs can be demonstrated in 20-30 lines of code!
//  - It doesn't use external libraries. These are often issues with the type definitions rather than TypeScript bugs.
//  - The incorrectness of the behavior is readily apparent from reading the sample.
// Reports are slower to investigate if:
//  - We have to pare too much extraneous code.
//  - We have to clone a large repo and validate that the problem isn't elsewhere.
//  - The sample is confusing or doesn't clearly demonstrate what's wrong.
namespace Case1 {
  interface Fnumber {(): number}
  interface Fstring {(): string}
  interface Fns {(): number|string}

  type F = Fns & Fnumber
  function test(f: F): ReturnType<F> {
    return f()
  }
}
```

### 🙁 Actual behavior

```
Type 'string | number' is not assignable to type 'number'.
  Type 'string' is not assignable to type 'number'.(2322)
(parameter) f: () => string | number (+1 overload)
```

### 🙂 Expected behavior

```
(parameter) f: () => number
```
The intersection type `F` should only have a return type of `number`

**Comment by caikan** [comment]
I found a workaround:
```ts
type NarrowReturn<F extends (...args: any) => any, R = any> =
  (ReturnType<F> & R) extends never ? never : (...args: Parameters<F>) => (ReturnType<F> & R)
```

**Comment by caikan** [comment]
@fatcerberus thank you for the explanation.
It looks like I had some misunderstandings about intersection types before.
Let me share my latest understanding:
Function call signatures (https://www.typescriptlang.org/docs/handbook/2/functions.html#call-signatures) are a special type of property in object types, and an object type can contain multiple call signatures. When intersecting object types, different call signatures are treated as different object properties. So when intersecting multiple different call signatures, we get a new object type that contains multiple call signatures, rather than a single "narrower" function type.

**Comment by fatcerberus** [comment]
Yep, that’s a really good way to think about it.  No attempt is made to merge the call signatures though, unlike with normal properties.
